p2p/protocol/circuitv2/relay/relay.go

```diff
@@ -164,30 +172,45 @@ func (r *Relay) handleStream(s network.Stream) {
 func (r *Relay) handleReserve(s network.Stream) {
 	defer s.Close()
 
+	if r.metricsTracer != nil {
+		r.metricsTracer.ReservationRequestReceived()
+	}
+
 	p := s.Conn().RemotePeer()
 	a := s.Conn().RemoteMultiaddr()
 
 	if isRelayAddr(a) {
 		log.Debugf("refusing relay reservation for %s; reservation attempt over relay connection")
-		r.handleError(s, pbv2.Status_PERMISSION_DENIED)
+		r.handleErrorAndTrackMetrics(s, pbv2.HopMessage_RESERVE, pbv2.Status_PERMISSION_DENIED,
```

Comment:
I'm not a big fan of the `handleError` function. I know you didn't introduce it, so this is not your fault :)

What about return an error and / or a status code, and handle it in the caller?

Reply:
lemme see how that looks.

Reply:
this looks much better! thanks!
I've left handleError as it is, if you want to remove that I can open a separate PR.

Reply:
Yeah, let's do this in a follow-up PR.